components/sup/src/manager/spec_watcher.rs

```diff
@@ -12,659 +12,298 @@
+//! services in response to the various `hab svc` commands.
+
+use std::{
+    num::ParseIntError,
+    str::FromStr,
+    sync::mpsc::{channel, Receiver},
+    thread::Builder,
+    time::Duration,
+};
+
+use notify::{DebouncedEvent, RecommendedWatcher, RecursiveMode, Watcher};
+
+use super::spec_dir::SpecDir;
+use config::EnvConfig;
 use error::{Error, Result};
-use manager::service::ServiceSpec;
 
 static LOGKEY: &'static str = "SW";
-const WATCHER_DELAY_MS: u64 = 2_000;
-const SPEC_FILE_EXT: &'static str = "spec";
-const SPEC_FILE_GLOB: &'static str = "*.spec";
-
-#[derive(Debug, PartialEq)]
-pub enum SpecWatcherEvent {
-    AddService(ServiceSpec),
-    RemoveService(ServiceSpec),
-}
 
-pub struct SpecWatcher {
-    watch_path: PathBuf,
-    have_events: Arc<AtomicBool>,
-}
-
-impl SpecWatcher {
-    pub fn run<P>(path: P) -> Result<Self>
-    where
-        P: Into<PathBuf>,
-    {
-        Self::run_with::<RecommendedWatcher, _>(path)
+/// How long should we wait to consolidate filesystem events?
+///
+/// This should strike a balance between responsiveness and
+/// too-granular a series of events.
+///
+/// See https://docs.rs/notify/4.0.6/notify/trait.Watcher.html#tymethod.new
+struct SpecWatcherDelay(Duration);
+
+impl From<Duration> for SpecWatcherDelay {
+    fn from(d: Duration) -> SpecWatcherDelay {
+        SpecWatcherDelay(d)
     }
+}
 
-    pub fn spec_files<T>(watch_path: T) -> Result<Vec<PathBuf>>
-    where
-        T: AsRef<Path>,
-    {
-        Ok(glob(
-            &watch_path
-                .as_ref()
-                .join(SPEC_FILE_GLOB)
-                .display()
-                .to_string(),
-        )?.filter_map(|p| p.ok())
-        .filter(|p| p.is_file())
-        .collect())
+impl Default for SpecWatcherDelay {
+    fn default() -> Self {
+        // There's nothing particularly magical about 2000ms,
+        // particularly since we're monitoring at such a coarse level
+        // ("something happened in this directory").
+        //
+        // Smaller is probably fine, but you wouldn't want to go much
+        // higher, as this could extend the amount of time you'd need
+        // to wait before realizing you need to take action on a
+        // service.
+        Duration::from_millis(2_000).into()
     }
+}
 
-    pub fn initial_events(&mut self) -> Result<Vec<SpecWatcherEvent>> {
-        self.generate_events(HashMap::new())
-    }
-
-    pub fn new_events(
-        &mut self,
-        active_specs: HashMap<String, ServiceSpec>,
-    ) -> Result<Vec<SpecWatcherEvent>> {
-        if self.have_fs_events() {
-            self.generate_events(active_specs)
-        } else {
-            Ok(vec![])
-        }
+impl FromStr for SpecWatcherDelay {
+    type Err = ParseIntError;
+    fn from_str(s: &str) -> ::std::result::Result<Self, Self::Err> {
+        // u16 ~= 65 seconds, which is _more_ than enough
+        let raw = s.parse::<u16>()?;
+        Ok(Duration::from_millis(raw as u64).into())
```

Comment:
That's fair... Really, no users should be messing with this anyway.

This does make me want to go back and find a decent way to denote at the code level that one of these `EnvConfig` things is meant to be "undocumented" and purely for internal use.

Reply:
I understand the desire, but I wonder if the difference between an an env var we expect users to use and one we don't really exists outside the code itself. There's no *behavior* that seems to change in the code. If users *do* start using this and we find enough of them set it to values that cause problems that make for a bad user experience, we can add more logic to help them from foot-shooting, but nothing we put in the code is likely to change whether any of these particular env vars gets used or not.

I think there's an important calculation about which we put in our documentation, since I think once that happens, we're obligated to continue supporting it or go through a formal deprecation procedure. But I don't think we should put any of them in our public-facing docs until experience shows they're useful and user-friendly.